runtime-sdk/src/modules/rofl/app/processor.rs

```diff
@@ -0,0 +1,186 @@
+
+        // Provision keys.
+        let signer = secp256k1::MemorySigner::random(&mut OsRng).unwrap();
+
+        // Create the processor task and start it.
+        let processor = Self {
+            state: Arc::new(State {
+                identity: state.identity.clone(),
+                host: state.protocol.clone(),
+                consensus_verifier: state.consensus_verifier.clone(),
+                signer: Arc::new(signer),
+                app: Arc::new(app),
+            }),
+            cmdq: rx,
+            cmdq_tx: tx.downgrade(),
+            logger: get_logger("modules/rofl/app"),
+            latest_round: 0,
+        };
+        tokio::spawn(processor.run());
+
+        tx
+    }
+
+    /// Run the processor task. This should be started inside a new task.
+    async fn run(mut self) {
+        slog::info!(self.logger, "starting processor";
+            "app_id" => A::app_id(),
+        );
+
+        // Prepare application environment.
+        let env = Environment::new(self.state.clone(), self.cmdq_tx.clone());
+
+        // Start subtasks.
+        let subtasks = Subtasks {
+            registration: super::registration::Task::start(self.state.clone(), env.clone()),
+            notifier: super::notifier::Task::start(self.state.clone(), env),
+        };
+
+        // Register for notifications.
+        let _ = self
+            .state
+            .host
+            .register_notify(host::RegisterNotifyOpts {
+                runtime_block: true,
+                runtime_event: vec![],
+            })
+            .await;
+
+        slog::info!(self.logger, "entering processor loop");
+        while let Some(cmd) = self.cmdq.recv().await {
+            if let Err(err) = self.process(&subtasks, cmd).await {
+                slog::error!(self.logger, "failed to process command";
+                    "err" => ?err,
+                );
+            }
+        }
+
+        slog::info!(self.logger, "processor stopped");
+    }
+
+    /// Process a command.
+    async fn process(&mut self, subtasks: &Subtasks, cmd: Command) -> Result<()> {
+        match cmd {
+            Command::ProcessRuntimeBlock(blk) => {
+                self.cmd_process_runtime_block(subtasks, blk).await
+            }
+            Command::GetLatestRound(ch) => self.cmd_get_latest_round(ch).await,
+            Command::InitialRegistrationCompleted => {
+                self.cmd_initial_registration_completed(subtasks).await
+            }
+        }
+    }
+
+    async fn cmd_process_runtime_block(
+        &mut self,
+        subtasks: &Subtasks,
+        blk: roothash::AnnotatedBlock,
+    ) -> Result<()> {
+        // Update latest known round.
+        if blk.block.header.round <= self.latest_round {
+            return Err(anyhow!("round seems to have moved backwards"));
+        }
+        self.latest_round = blk.block.header.round;
+
+        // Notify registration task.
+        let _ = subtasks.registration.try_send(());
```

Comment:
> Maybe we could hide tx, store the task instead, and call subtasks.registration.refresh();. Same below, so that you don't need to import Notify enum.

That could work but similar as in my other comment it would require splitting the task into implementation and handle, then the handle would have the refresh method.